src/UI/class-network-admin-sites-list.php

```diff
@@ -0,0 +1,110 @@
+ *
+ * @package Parsely
+ * @since 3.2.0
+ */
+
+declare(strict_types=1);
+
+namespace Parsely;
+
+use Parsely\Parsely;
+
+/**
+ * Render the additions to the WordPress Multisite Network Admin Sites List page
+ *
+ * @since 3.2.0
+ */
+final class Network_Admin_Sites_List {
+	/**
+	 * Constructor.
+	 *
+	 * @param Parsely $parsely Instance of Parsely class.
+	 */
+	public function __construct( Parsely $parsely ) {
+		$this->parsely = $parsely;
+	}
+
+	/**
+	 * Attach network admin page functionality to the appropriate action and filter hooks.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return void
+	 */
+	public function run() {
+		add_filter( 'manage_sites_action_links', array( __CLASS__, 'add_action_link' ), 10, 2 );
+		add_filter( 'wpmu_blogs_columns', array( __CLASS__, 'add_api_key_column' ) );
+		add_action( 'manage_sites_custom_column', array( $this, 'populate_api_key_column' ), 10, 2 );
+	}
+
+	/**
+	 * Use the manage_sites_action_links filter to append a link to the settings page in the "row actions."
+	 *
+	 * @param array $actions The list of actions meant to be displayed for the current site's context in the row actions.
+	 * @param int   $_blog_id The blog ID for the current context.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return array The list of actions including ours.
+	 */
+	public static function add_action_link( array $actions, int $_blog_id ): array {
+		if ( ! current_user_can( Parsely::CAPABILITY ) ) {
+			return $actions;
+		}
+
+		if ( ! is_numeric( $_blog_id ) ) {
+			return $actions;
+		}
+
+		$settings_url                = Parsely::get_settings_url( (int) $_blog_id );
+		$actions['parsely-settings'] = '<a href="' . esc_url( $settings_url ) . '">' .
+		__( 'Parse.ly Settings', 'wp-parsely' ) .
+		'</a>';
+
+		return $actions;
+	}
+
+	/**
+	 * Use the wpmu_blogs_columns filter to register the column where we'll display the site's API Key (if configured).
+	 *
+	 * @param array $sites_columns The list of columns meant to be displayed in the sites list table.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return array The list of columns to display in the network admin table including ours.
+	 */
+	public static function add_api_key_column( array $sites_columns ): array {
+		$sites_columns['parsely-api-key'] = __( 'Parse.ly API Key', 'wp-parsely' );
+		return $sites_columns;
+	}
+
+	/**
+	 * Use the manage_sites_custom_column action to output each site's API Key (if configured).
+	 *
+	 * @param string $column_name The column name for the current context.
+	 * @param int    $_blog_id The blog ID for the current context.
+	 *
+	 * @since 3.2.0
+	 *
+	 * @return void
+	 */
+	public function populate_api_key_column( string $column_name, int $_blog_id ): void {
+		if ( 'parsely-api-key' !== $column_name ) {
+			return;
+		}
+
+		// phpcs:ignore WordPressVIPMinimum.Functions.RestrictedFunctions.switch_to_blog_switch_to_blog
+		switch_to_blog( $_blog_id );
+		$apikey = $this->parsely->get_api_key();
+		restore_current_blog();
```

Comment:
If a site has 100 or 4000 sites, how painful will this bit of code be? Does the database retrieval across all of the options tables for the subsites need to be cached in the object cache, or stored in an option in the root site, if say, the number of subsites is above a certain threshold?

Reply:
> If a site has 100 or 4000 sites, how painful will this bit of code be? 

Short answer is:
It's probably ok.

Here's some context:

1. The list table is [paginated](https://github.com/WordPress/WordPress/blob/270f2011f8ec7265c3f4ddce39c77ef5b496ed1c/wp-admin/network/sites.php#L17-L18), so I think it will be capped at 100 sites at a time (the [default](https://github.com/WordPress/WordPress/blob/fa62df07749a3bfaa44e944c02512215ab70a5db/wp-includes/class-wp-site-query.php#L127) for a `WP_Site_Query`).
1. `Parsely::get_api_key` is a pretty lightweight wrapper around `get_option`, which is [backed by an object cache](https://github.com/WordPress/WordPress/blob/2a3ad7c4304143c4761726c7c49beda995d6f2f0/wp-includes/option.php#L171).
1. `get_option` [calls](https://github.com/WordPress/WordPress/blob/2a3ad7c4304143c4761726c7c49beda995d6f2f0/wp-includes/option.php#L166) `wp_load_alloptions` prior to the lookup. That result is also [backed](https://github.com/WordPress/WordPress/blob/2a3ad7c4304143c4761726c7c49beda995d6f2f0/wp-includes/option.php#L279-L280) by object cache. Our option is set to autoload by default, so, this is the codepath that will provide the data in most cases when the option is set for a subsite.
1. For subsites without the option set, a [query](https://github.com/WordPress/WordPress/blob/2a3ad7c4304143c4761726c7c49beda995d6f2f0/wp-includes/option.php#L174) will be executed the first time the option is attempted to be read for each subsite.  When the key is not found, `get_option` stores that fact in [another cache](https://github.com/WordPress/WordPress/blob/2a3ad7c4304143c4761726c7c49beda995d6f2f0/wp-includes/option.php#L146-L186), so this query will not happen on subsequent calls to `get_option`.
1. The option query linked above is super fast thanks to a `UNIQUE KEY` on `wp_#_options`.`option_name`, so 100 times shouldn't be a huge deal. The bottleneck is likely the IO between the web server & db as opposed to any processing load on the db.
1. The option will be populated and cache hot the very first time a subsite is accessed in any manner owing to [the way `Parsely::run` works](https://github.com/Parsely/wp-parsely/blob/f536ee6e61a738a4843e2b7aa532c0321b440211/src/class-parsely.php#L98-L109)

All that to say, I don't think this will have a meaningful impact on site performance -- particularly on a production site with persistent caching.

I'd say the largest impact we could expect on a large multisite would be while network-activating the plugin. The first time a subsite is accessed in any manner afterward, the option will be written and cache populated.